File: PDOO-Napakalaki/Napakalaki/NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "bad_consequence"

class SpecificBadConsequence < BadConsequence
  
  attr_accessor :someSpecificVisibleTreasures, :someSpecificHiddenTreasures
  
  def initialize(aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    super(aText, someLevels)
    @someSpecificVisibleTreasures = someSpecificVisibleTreasures
    @someSpecificHiddenTreasures = someSpecificHiddenTreasures
  end
  
  def isEmpty
    empty = false
        
    if(@someSpecificVisibleTreasures.empty? && @someSpecificHiddenTreasures.empty?)
      empty = true
    end 
        
    return empty
  end
  
  
  def substractVisibleTreasure(t)      
    if(!(@someSpecificVisibleTreasures.empty?))
      @someSpecificVisibleTreasures.delete(t.treasureKind)
    end
        
  end
    
   
  def substractHiddenTreasure(t)     
    if(!(@someSpecificHiddenTreasures.empty?))
      @someSpecificHiddenTreasures.delete(t.treasureKind)
    end
  end
    
  def adjustToFitTreasureLists(v, h)          
        
        tVisibleT = Array.new
        tHiddenT = Array.new  
        tVisible = Array.new
        tHidden = Array.new
        bc = SpecificBadConsequence.new("", 0, Array.new, Array.new)
       
        if(@someSpecificHiddenTreasures == nil) 
          @someSpecificHiddenTreasures = Array.new
        end
        if(@someSpecificVisibleTreasures == nil) 
          @someSpecificVisibleTreasures = Array.new
        end
              
        if(!(@someSpecificVisibleTreasures.empty?) || !(@someSpecificHiddenTreasures.empty?))
                 
          v.each { |t|
            tVisibleT << t.treasureKind
          }
          
          h.each { |t|
            tHiddenT << t.treasureKind
          }
          
          @someSpecificVisibleTreasures.each { |t|
            if(tVisibleT.index(t) != nil)
              tVisible << t
              tVisibleT.delete_at(tVisibleT.index(t))
            end
          }
          
          @someSpecificHiddenTreasures.each { |t|
            if(tHiddenT.index(t) != nil)
              tHidden << t
              tHiddenT.delete_at(tHiddenT.index(t))
            end
          }         
          bc = SpecificBadConsequence.new("", 0, tVisible, tHidden)         
        end        
        
        return bc
  end
end

File: PDOO-Napakalaki/Napakalaki/NapakalakiJava/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author ramonjtorres
 */
public class SpecificBadConsequence extends BadConsequence{
    
    
    private ArrayList<TreasureKind>specificVisibleTreasures=new ArrayList();
    private ArrayList<TreasureKind>specificHiddenTreasures=new ArrayList();
    
    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden) {
        super(levels,text);
        this.specificHiddenTreasures=tHidden;
        this.specificVisibleTreasures=tVisible;
    }
        
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    @Override
    public  boolean isEmpty(){
        
        boolean devolver=false;
        
        if(!specificVisibleTreasures.isEmpty()&&!specificHiddenTreasures.isEmpty())
            devolver=true;
        
        return devolver;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureList(ArrayList<Treasure>v,ArrayList<Treasure>h){
        
        BadConsequence devolver;
        ArrayList<TreasureKind>sVis=new ArrayList();
        ArrayList<TreasureKind>sHid = new ArrayList();

        if(this.specificVisibleTreasures!=null)
            sVis= specificVisibleTreasures;
            
        if(this.specificHiddenTreasures!=null)
            sHid = this.specificHiddenTreasures;
            
        ArrayList<TreasureKind>aux1 = new ArrayList();
        ArrayList<TreasureKind>aux2 = new ArrayList();

            
        for(Treasure tmp : v){
            if( sVis.remove(tmp.getType()) )
               aux1.add(tmp.getType());
        }
            
        for(Treasure tmp : h){
            if( sHid.remove(tmp.getType()) )
               aux2.add(tmp.getType());
        }
            
        devolver=new SpecificBadConsequence(getText(), getLevels(), aux1, aux2);
        return devolver;
    }

    @Override
    public void substractVisibleTreasure(Treasure t){
        
        if(!this.specificVisibleTreasures.isEmpty())
            this.specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        
        if(!this.specificHiddenTreasures.isEmpty())
            this.specificHiddenTreasures.remove(t.getType());
    }

    @Override
    public String toString() {
        return super.toString() + "specificVisibleTreasures=" + specificVisibleTreasures + ", specificHiddenTreasures=" + specificHiddenTreasures + '}';
    }

    @Override
    public boolean isDead() {
        return false;
    }
    
}
